### PR TITLE
[Agent] (Potentially) improve Editing using `diff`

### PR DESCRIPTION
**Pull request by xingyaoww**
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**

Currently, agents sometimes struggle with `edit_file` skill that requires them to input line numbers - this is challenging to current foundation models, even the strongest (e.g., instance `astropy__astropy-14995` using `CodeActAgent v1.6-no-hint` and `gpt-4o` - check visualizer for entire history https://huggingface.co/spaces/OpenDevin/evaluation).

The agent generated legit code to fix the issue but accidentally inserted the code into a docstring and triggered a syntax error.
image

**Give a brief summary of what the PR does, explaining any non-trivial design decisions**

This PR implements `diff`-based similar to [aider](https://aider.chat/docs/benchmarks.html#diff) and resolves https://github.com/OpenDevin/OpenDevin/issues/2222.

TODO: run the implementation on SWE-Bench-Lite to see if it improves score.

**Other references**

### CODE REVIEW

**Review by tobitege — CHANGES_REQUESTED** [review]
The "replace first in file" approach should be implemented independently of the existing methods as those cover their own cases.
I hope my above comments explain why I'm a little concerned with this.

**Comment by xingyaoww** [comment]
> The "replace first in file" approach should be implemented independently of the existing methods as those cover their own cases.

We are aiming for an evaluation-driven approach: We will only consider merging this if this alone improves the SWE-Bench score; If it degrades the score in any way, we definitely won't merge it.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (xingyaoww) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:336]: > changing this method this way will cause "fixed" issues to reappear.

Did we have a test for this "fixed" issue though? 

>  that a single line edit was adding an extra blank line together with the content inside of files.

Why was this an issue though?

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (xingyaoww) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:518]: This is actually what aider's edit block does - which we are interested in benchmarking against - let's wait for the score before we decide which is better :D

https://github.com/OpenDevin/OpenDevin/issues/2222#issuecomment-2197434573

https://github.com/paul-gauthier/aider/blob/4753db0b0f6b597410fda1288d484f367819fed5/aider/coders/editblock_coder.py#L61-L64

File: tests/integration/mock/CodeActAgent/test_browse_internet/prompt_001.log
Comment (xingyaoww) [reply tests/integration/mock/CodeActAgent/test_browse_internet/prompt_001.log:365]: This is certainly more like a bug rather than feature which i hope to get rid of - right now I'm just hoping to run evaluation on this to see if it has the potential to replace current method. Dang i should mark this as draft

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (xingyaoww) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:336]: Oh nvm - i saw the follow up comment!

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (xingyaoww) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:513]: > Imo this raises the chances of an accidental replacement at the wrong occurence as this has to work with any kind of text, not just with files from eval data, that might be "unique".

Definitely a possibility - maybe it is just a question of "how much". 

Ideally, I very much like the line-based editing approach because it is clean, complete, and doesn't involve too much heuristics.

But these line-based edits are mainly bottlenecked by the current LLM's capability - things may improve in a year, but in the meanwhile, aside from the training model (which we are already thinking & doing), exploring different editing approaches like this could be another good alternative in improving the agent quality.

Comment (tobitege) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:513]: I'm curious about your screenshot: did the LLM edit that same file already before?
That could explain that it itself doesn't keep track of line numbers and didn't care to re-read the changed file before applying an edit "blindly" the 2nd time.
This could be avoided by a slightly changed prompt, potentially, by advising the LLM to actually refresh its context about the location _before_ it wants to edit again.
Even in my use of LLM's as assistants in my IDE (Cursor), I have to double-check every code suggestion as it constantly makes assumptions and with only a small "task" in mind, looses the big picture, and ignores consequences from related issues it may cause... but I digress. 😂

Comment (xingyaoww) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:513]: > did the LLM edit that same file already before?

Yes it did!

> This could be avoided by a slightly changed prompt, potentially, by advising the LLM to actually refresh its context about the location before it wants to edit again.

Hmm, I think this is also possible! But still, the other possibility is that this is too challenging for today's LLM to figure out "line number" - Let's wait for score haha 😂

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (tobitege) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:518]: > This is actually what aider's edit block does - which we are interested in benchmarking against - let's wait for the score before we decide which is better :D

Let me play devil's advocate for a minute here. I understand the use case for research and benchmarks.
But OpenDevin should not be only tuned for that, it may fail in every other real-world scenario where file sizes and languages differ wildly. We may have different perspectives on this, so I hope you don't mind me looking at it differently, too. 🤗

Comment (xingyaoww) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:518]: > OpenDevin should not be only tuned for that... it may fail in every other real-world scenario where file sizes and languages differ wildly.

Definitely agree on this - I think very hard about how to make agents excel at software tasks (e.g., issue fixing) while also being general enough - that's why we have a very diverse array of benchmarks in OpenDevin now.

But my take is that before we have benchmark / real qualitative evidence that "it may fail in every other real-world scenario," SWE-Bench benchmark is somewhat a good "middle-ground" for us to estimate its practical usefulness.

Plus, if this performs well, "Make a text range "uniqe enough" seems very indeterministic." is not unfixable - just takes more effort!

**If** this PR is actually going to give us better score, I think we can probably work together to figure out a way to make both of us happy :D Just need sometime to discuss and brainstorm!

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (tobitege) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:513]: It's more from what I experienced in using OpenDevin where there were these cases, as you observed, where it made multiple edits _without_ checking first. 
I'd _hope_ LLM's would get trained to act accordingly, but for now I'd just make more precise instructions for it to follow.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (tobitege) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:518]: Happy to, just still don't see why it shouldn't be its standalone method. 😬

Comment (xingyaoww) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:518]: >  it may fail in every other real-world scenario where file sizes and languages differ wildly.

It just occurs to me - since there's no benchmark testing this "editing capability" across different programming languages (to the best of my knowledge ofc) and use cases, why can't we, as a community, build one that reflects our *expectation* to what type of *real-world* edit OpenDevin we hope should be able to make? And we can use that benchmark to compare different editing tools - and potentially make it a publishable academic paper!

The **goal** is to build a comprehensive evaluation benchmark to **measure how well an agent can perform *edits*** -  it should try to cover as many OpenDevin use cases as possible. This will guide us towards developing better `agentskills` for editing.

**Why not existing benchmarks?** SWE-Bench is a very challenging task that requires the agent to (1) search the repo, (2) find a place to edit, (3) generate correct code, and (4) perform edits. That means its performance is hard to interpret as to what exactly went wrong. On the other hand, we do have existing editing benchmarks like [aider](https://aider.chat/docs/benchmarks.html#diff); however, as @tobitege suggested, this is mainly based on Python and cannot easily cover most people's use case of OpenDevin (ppl work on different programming languages!).

Ideally, in this benchmark, we should just measure the ability to *(4) perform edit* assume we know what we want the agent to edit; and the performance we measure should not be confounded with the model's ability to *generate code*. 

Maybe we can use **a Github commit diff patch** as **input** to the model - we then ask the model to use whatever tools it has (e.g., different implementation of `agentskills`) and edit the current repository/file to reproduce the git diff patch. To measure its success, when the edit is completed, we measure the `git diff` of the edited repository and compare it with the initial given `git diff` patch.

This may sound easy, but I suspect a lot of LLM will make mistakes easily, especially when running into less popular programming languages.

For an executable plan, my thought is that:

1. **Data collection:** We can start with Github commits across different languages: A ["commitpack" dataset](https://huggingface.co/datasets/bigcode/commitpack) contains 4TB github data across 350 programming languages. We can sample a subset of each language and get the repository state before editing, as well as the commit patch.
2. **Run evaluation**: Similar to what we did with SWE-Bench eval now, we can just set up the environment for an agent (e.g., `swe_env_box.py`) and give it the `git patch` as part of the instruction and let it do the rest!
3. **Use this eval to improve OpenDevin's editing performance**: Implement different `agentskills` and evaluate them!
4. (Optionally) We can collect another subset, ask the agent to perform edits, keep the successful trajectories, and use them to train open-source models so they are better at editing!

Some caveats:
- Even if "commitpack" is likely already used to train LLM, I suspect it won't make a difference: Even if the agent remembers all the code, it is still possible to fail at editing them.
- If that is indeed a concern, we can easily crawl some of the most recent (e.g., since 2024 May) github commits across different languages and make them a dataset!

Comment (neubig) [reply opendevin/runtime/plugins/agent_skills/agentskills.py]: Hey @tobitege , I just read through the discussion here.

* First, I think this is a great discussion!
* Second, I think that it might be worth keeping around the current agentskills, even if after this PR they are not used by CodeAct. That would allow us to do experiments in the future.
* Third, in this particular case I think we shouldn't let perfect be the enemy of progress. SWE-bench is the best benchmark that we have at the moment, and creating a new benchmark like the one @xingyaoww mentioned would take a month at least and lots of effort.

So to summarize, my suggestion would be (if it's not too onerous) to keep "append_file" in the code, but not enable it for CodeAct 1.7.

Some other context is that I'm currently running SWE-bench with the new code and it seems to be working reasonably well.

Comment (tobitege) [reply opendevin/runtime/plugins/agent_skills/agentskills.py]: I fear that might have an adverse effect, because append is a specific use case to supplement the edit_file method: per the logic of it, you have to replace something, but that basically excludes appending, because you can't replace "line + 1" which wasn't allowed and the LLM would have to replace the "last line + new content".
Looking forward to see this progress, just wanted to clarify why I see append as beneficial in any case. 🤗

Comment (neubig) [reply opendevin/runtime/plugins/agent_skills/agentskills.py]: Yes, I totally understand!

In terms of sequencing, I'd suggest that I finish the eval (should be done today). If it continues to look good, the eval was run exactly on this branch, so we will know that this branch increases our SWE-bench scores, and because that is currently one of our major goals we go ahead and merge this in.

Then after we do that, we create another branch with "append" enabled and evaluate that. If that further improves scores then we could add it back.

To be clear, this is more a practical concern of evaluation being expensive, both from the perspective of time and cost, so once we have positive results it'd probably be good to get them merged in.

Separately from this discussion, I understand that SWE-bench is not perfect and not reflective of all use cases, so having other benchmarks is also a great idea!

**Comment by neubig** [comment]
This PR is currently blocked by me, I need to finish evaluation, hopefully today.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (xingyaoww) [reply opendevin/runtime/plugins/agent_skills/agentskills.py]: I think `append` may be somewhat related to `insert_content_at_line`, though it will have the assumption of "it needs a newline at the end of the file." I agree with @neubig that we could test this in a next PR, and maybe merge this one as is (if it improve performance)? It won't affect user experience too much now since it is on `main`, not on a released version.

Comment (xingyaoww) [reply opendevin/runtime/plugins/agent_skills/agentskills.py]: I started a draft PR where we can do this:
https://github.com/OpenDevin/OpenDevin/pull/2722

**Review by neubig — APPROVED** [review]
I evaluated 1.7 and here are the results (all no hints):

- 1.6 w/ gpt-4o: 19.8
- 1.7 w/ gpt-4o: 21.33
- 1.7 w/ claude-3.5-sonnet: 24.33

So it seems like the current state as represented by this PR has made some improvement over 1.6 when I previously evaluated. Based on this I think we can go ahead with this PR!

**Comment by neubig** [comment]
@tobitege this is blocked by your "request changes" if you could please take a look I think it's ready other than that!

**Review by tobitege — APPROVED** [review]

**Comment by tobitege** [comment]
> @tobitege this is blocked by your "request changes" if you could please take a look I think it's ready other than that!

Done now.
@xingyaoww so the extra newline is still there (see tests). Might keep an eye out for that.

**Comment by xingyaoww** [comment]
@tobitege I already fixed that in https://github.com/OpenDevin/OpenDevin/pull/2722! I'll try to get this one ready and evaluated soon!

**Comment by tobitege** [comment]
Oh, got it. I lost track 😁

**Pull request merged**
